### PR TITLE
Feature | join section

### LINKED ISSUE #280: Join Section User interaction [Design document]
## Task Flow:

- 1. User visits the main site https://www.realdevsquad.com 
        - Since the feature is still under development we have to pass a query parameter `join=true`
- 2. The user finds a section to join with a button
- 3. On clicking the button the user is redirected to the join page of the Realdevsquad website.
- 4. Here the user has to fill in their details (Handled using my site feature being built by Vinayak Trivedi)
- 5. On completion of the flow of my site user is redirected to a form where they fill in RDS Related details
        - a. Introduction
        - b.  How many hours can they commit to RDS
        - c.  Why do they want to join RDS
        - d. Others (More fields could be added based on further discussion)
- 6. On completion of this form, they will get a link, which needs to be sent for verification (currently shared manually to Ankush)
        - 6.1. Once the form details are verified they will get a fresh discord link to join the server (sent manually by Ankush)
- 7. On the final page, the user will only see the link to share and a message to wait for the link if the details get verified

### LINKED ISSUE #286: [RFC] Storage of join section data
## Problem Description
Once the user enters his/her details in the join section a personalized link is created for them to share with the verifier. This link contains their data. The **Issue** here is whether to store the data after verification of the user or as soon as the user submits the form.

## Current flow
- user signs in using the GitHub login.
- redirected to new signup flow by @vinayak-trivedi 
- after the signup flow user is sent to the join section to enter the details.
- a user-specific link is generated which has his/her data.
- link shared to the verifier to verify and send the discord invite link.

**Note:** The issue is to store the data before verification or after verification.

### PR DESCRIPTION
**Pull request by RitikJaiswal75**
**Will be functional after https://github.com/Real-Dev-Squad/website-backend/issues/714 is closed**
**Link for the back-end PR :** https://github.com/Real-Dev-Squad/website-backend/pull/715
## Problem Statement
Once the User Signs in to Peal Dev Squad, He / She has to wait for the discord invite of the real dev squad. Currently, this process is done by Ankush through manual acquisition and takes a lot of time asking the same question to everyone who wants to join Real Dev Squad.

## Solution Discussed
Create a Join section for Real dev Squad where the questions are already mentioned, and the user answers them. Once they answer all the questions, generate a personalized link for them that contains the data they entered.

## Current status:
- [x] Front-end part of the collection and sending data to the backend
- [x] Backend Task to upload and fetch data 

## References to the solution Discussions:
- https://github.com/Real-Dev-Squad/website-www/issues/280
- https://github.com/Real-Dev-Squad/website-www/issues/286

## Points to reviewers:
- the Percy generate snapshots test will fail as I have made UI changes. That cannot be fixed currently.
- In case any other test fails or you find any issues with file changes please comment here.

## Expected Flow
- User comes to [Real Dev Squad website](httpw://realdevsquad.com).
- They click on sign in with GitHub.
- complete the new/old signup flow (Whichever is active at the time.)
- User now has some data on Real Dev Squad
- Now we ask the user to answer the questions that were earlier asked manually with the help of this join section
  - User is sent to location `realdevsquad.com/join.html`
  - Page 1: Start the join process
  - Page 2: Enter personal (Location) Details
  - Page 3: Enter their introduction
  - page 4: why do they want to join RDS and how did they hear about us
  - page 5: Preview of the data Enter
  - page 6: final page with next steps and a link that contains personal info.
- Now the user waits for the data to be verified and after which they get the joining link for a discord server. 

### Issue
Closes #280 
Closes #286 

### What is the change?

Making join section for user introduction and verification

### Is Development Tested?

- [x] Yes
- [ ] No

https://user-images.githubusercontent.com/57758447/192142540-dafc63b0-773f-4d00-8744-0a5505661167.mp4

### CODE REVIEW

File: join.html [hunk 1]
```diff
@@ -149,31 +156,244 @@
...
+          <input
+            class="user-input input-regular"
+            type="text"
+            name="state"
+            id="state"
+            placeholder="Enter your state"
+          />
+          <br />
+          <label class="input-label" for="country">
+            Your country
```
Comment (rohan09-raj) [inline join.html:201]: Can we please be consistent with the naming?

Comment (rohan09-raj) [reply join.html:201]: country -> Country

File: join.html [hunk 2]
```diff
@@ -149,31 +156,244 @@
...
+            placeholder="Enter your country"
+          />
+          <div class="twin-button-container">
+            <button class="button-outline" id="previous1">Previous</button>
+            <button class="button-disabled nextBtn" id="next1">Next</button>
+          </div>
+        </div>
+      </section>
+      <section class="page3" id="page3">
+        <h1 class="greeting">Let’s Help Others 🫂 know you and your skills.</h1>
```
Comment (rohan09-raj) [inline join.html:219]: Help Others -> help others

File: join.html [hunk 3]
```diff
@@ -149,31 +156,244 @@
...
+          <p class="input-label-dark">How did you hear about us</p>
+          <p class="user-input input-regular" id="previewHeardAbout"></p>
+          <div class="twin-button-container">
+            <button class="button-outline" id="previous4">Previous</button>
+            <button class="button-filled" id="next4">Submit</button>
+          </div>
+        </div>
+      </section>
+      <section class="page6" id="page6">
+        <h1 class="greeting">Thankyou 🫂 for completing all the steps.</h1>
```
Comment (rohan09-raj) [inline join.html:372]: Thankyou -> Thank you

File: join.html [hunk 4]
```diff
@@ -149,31 +156,244 @@
...
+          </p>
+          <p class="user-input-dark">Next Steps</p>
+          <ol>
+            <li>Copy the above link</li>
+            <li>
+              Follow Ankush Dharkar on
+              <a href="https://twitter.com/ankushdharkar">Twitter</a>
+            </li>
+            <li>
+              Message this link to him with a text of RDS Verification Link.
```
Comment (rohan09-raj) [inline join.html:387]: RDS -> Real Dev Squad

File: js/join.js [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      localStorage.setItem('flowState', (currentFlowState += 1));
+      previewFiller();
+    }
+    selectPage();
+  });
+});
+
+submit.addEventListener('click', async () => {
+  let data = JSON.stringify(localStorage);
+  let method = 'POST';
```
Comment (rohan09-raj) [inline js/join.js:273]: Can we please make it as `const`?

File: js/join.js [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+function showPage(currentFlowState) {
+  for (let i = 0; i < pages.length; i++) {
+    if (i == currentFlowState) {
+      pages[i].classList.remove('hide-page');
+    } else {
+      pages[i].classList.add('hide-page');
+    }
+  }
```
Comment (rohan09-raj) [inline js/join.js:123]: Can we please simplify this with a toggle? Also, we can use a forEach loop
For reference: https://replit.com/@rohan1914/Practice-2

File: js/join.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    next2.classList.remove('button-filled');
+  }
+  if (whyRdsPageValidator()) {
+    previewBtn.classList.remove('button-disabled');
+    previewBtn.classList.add('button-filled');
+  } else {
+    previewBtn.classList.add('button-disabled');
+    previewBtn.classList.remove('button-filled');
+  }
+}
```
Comment (rohan09-raj) [inline js/join.js:182]: I think we can simplify this with the toggle property.

File: js/join.js (hunk 6 above)
Comment (RitikJaiswal75) [reply js/join.js:123]: I just want only 1 page to appear at one time

File: js/join.js (hunk 7 above)
Comment (RitikJaiswal75) [reply js/join.js:182]: Each button serves a differnt functioality to work properly with toggles

File: css/join.css [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  font: 500 18px 'Roboto';
+  line-height: 21px;
+  margin: 0 0 2rem;
+}
+
+.info,
+.input-label,
+.input-label-dark {
+  font: 300 16px 'Roboto';
+  line-height: 19px;
```
Comment (rohan09-raj) [inline css/join.css:31]: Can we please be consistent with the units of line height?

Comment (rohan09-raj) [reply css/join.css:31]: Make the line height on line 18 in `px`

File: css/join.css [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+.input-label-dark {
+  font-weight: 500;
+}
+
+.user-input {
+  font: 300 16px 'Roboto';
+  background-color: var(--input-background);
+  width: 30rem;
+  line-height: 2rem;
```
Comment (rohan09-raj) [inline css/join.css:46]: Here too can we please use `px`

File: css/join.css [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  border: 1px solid var(--correct-border);
+}
+
+@media screen and (min-width: 611px) {
+  .user-input {
+    width: 35rem;
+  }
+  .user-input-big {
+    width: 35rem;
+  }
```
Comment (rohan09-raj) [inline css/join.css:134]: Can we please merge this?

File: css/join.css [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+  background-color: var(--error-background);
+  border: 1px solid var(--error-border);
+}
+
+.correct-data {
+  background-color: var(--correct-background);
+  border: 1px solid var(--correct-border);
+}
+
+@media screen and (min-width: 611px) {
```
Comment (rohan09-raj) [inline css/join.css:128]: Can we please use 625px here?
I am planning to use specific media queries to make responsive on all devices in the ember-migration

File: css/join.css [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+.container {
+  width: max-content;
+  margin: 0 auto;
+  max-width: 35rem;
+}
+
+.container ol li {
+  line-height: 2rem;
```
Comment (rohan09-raj) [inline css/join.css:18]: Please make it `px` to have consistency

File: css/join.css [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+.correct-data {
+  background-color: var(--correct-background);
+  border: 1px solid var(--correct-border);
+}
+
+@media screen and (min-width: 625px) {
+  .user-input,
+  .user-input-big {
+    width: 35rem;
+  }
```
Comment (rohan09-raj) [inline css/join.css:132]: Please have a line space after a CSS block

File: join.html [hunk 14]
```diff
@@ -149,31 +156,245 @@
...
+      </section>
+      <section class="page4" id="page4">
+        <h1 class="greeting">
+          Let’s help the verifier understand 🤔 you better.
+        </h1>
+        <div class="container">
+          <h2 class="progress">Step 3 of 4: Why RDS?</h2>
+          <p class="info">Fields marked with * are mandatory</p>
+          <label class="input-label" for="whyRds">
+            Why do you want to join RDS?
```
Comment (rohan09-raj) [inline join.html:299]: Please make it `Real Dev Squad`

File: join.html [hunk 15]
```diff
@@ -149,31 +156,245 @@
...
+            <button class="button-disabled nextBtn" id="next2">Next</button>
+          </div>
+        </div>
+      </section>
+      <section class="page4" id="page4">
+        <h1 class="greeting">
+          Let’s help the verifier understand 🤔 you better.
+        </h1>
+        <div class="container">
+          <h2 class="progress">Step 3 of 4: Why RDS?</h2>
```
Comment (rohan09-raj) [inline join.html:296]: RDS -> Real Dev Squad

File: join.html [hunk 16]
```diff
@@ -149,31 +156,245 @@
...
+          <br />
+          <select
+            name="heardAbout"
+            id="heardAbout"
+            class="user-input input-regular"
+          >
+            <option value="">Select an option</option>
+            <option value="Twitter">Twitter</option>
+            <option value="Linkedin">Linkedin</option>
+            <option value="Referred">Referred by S\someone from RDS</option>
```
Comment (rohan09-raj) [inline join.html:324]: 1. What is `S\someone`?
2. RDS -> Real Dev Squad

**Pull request merged**